static/app/stores/Commiters/CommittersContext.tsx

```diff
@@ -0,0 +1,279 @@
+  children: React.ReactNode;
+  initialState?: CommittersState;
+}
+
+export function CommittersProvider(props: CommittersContextProviderProps) {
+  const contextValue = useReducer(committersReducer, props.initialState ?? {});
+
+  return (
+    <CommittersContext.Provider value={contextValue}>
+      {props.children}
+    </CommittersContext.Provider>
+  );
+}
+
+export function useCommitters(): [CommittersState, React.Dispatch<CommittersAction>] {
+  const context = useContext(CommittersContext);
+
+  if (!context) {
+    throw new Error('useCommitters called outside of CommittersContext.Provider');
+  }
+
+  return context;
+}
+
+interface RequiredWithCommittersProps {
+  event: Event;
+  organization: Organization;
+  project: Project | AvatarProject;
+  group?: Group;
+}
+
+export interface WithCommittersProps {
+  committers: Committer[] | undefined;
+}
+
+export function withCommitters<P extends WithCommittersProps>(
+  Component: React.ComponentType<P>
+) {
+  const wrappedComponent: React.ComponentType<
+    Omit<P, keyof WithCommittersProps> & RequiredWithCommittersProps
+  > = (props): React.ReactElement => {
+    const api = useApi();
+    const [state, dispatch] = useCommitters();
+
+    useEffect(() => {
+      if (!props.group?.firstRelease) {
+        return undefined;
+      }
+
+      const key = makeCommitterStoreKey({
+        organizationSlug: props.organization.slug,
+        projectSlug: props.project.slug,
+        eventId: props.event.id,
+      });
+
+      if (state[key]?.committers || state[key]?.committersLoading) {
+        return undefined;
+      }
+
+      dispatch({
+        type: 'start loading',
+        payload: {
+          eventId: props.event.id,
+          organizationSlug: props.organization.slug,
+          projectSlug: props.project.slug,
+        },
+      });
+
+      let unmounted = false;
+
+      fetchCommitters(api, {
+        organizationSlug: props.organization.slug,
+        projectSlug: props.project.slug,
+        eventId: props.event.id,
+      })
+        .then(response => {
+          if (unmounted) {
+            return;
+          }
+
+          dispatch({
+            type: 'add committers',
+            payload: {
+              committers: response.committers,
+              eventId: props.event.id,
+              organizationSlug: props.organization.slug,
+              projectSlug: props.project.slug,
+            },
+          });
+        })
+        .catch(() => {
+          if (unmounted) {
+            return;
+          }
+          dispatch({
+            type: 'set error',
+            payload: {
+              eventId: props.event.id,
+              organizationSlug: props.organization.slug,
+              projectSlug: props.project.slug,
+            },
+          });
+        });
+
+      return () => {
+        unmounted = true;
+      };
+    }, [props.group, props.organization.slug, props.project.slug, props.event.id]);
+
+    const key = makeCommitterStoreKey({
```

Comment:
we have key here and inside of `useEffect`, shouldn't we just move this code to the very top so we can reuse it?

Reply:
I think I prefer to call it twice because of how hooks create closures, it's less error prone and more future proof to just recalculate it